sdk/ton/timelock_converter.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/smartcontractkit/mcms/types"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/xssnick/tonutils-go/address"
+	"github.com/xssnick/tonutils-go/tlb"
+	"github.com/xssnick/tonutils-go/tvm/cell"
+
+	"github.com/smartcontractkit/chainlink-ton/pkg/bindings/mcms/timelock"
+	"github.com/smartcontractkit/chainlink-ton/pkg/ton/tlbe"
+	"github.com/smartcontractkit/chainlink-ton/pkg/ton/tvm"
+)
+
+var _ sdk.TimelockConverter = (*timelockConverter)(nil)
+
+type timelockConverter struct{}
+
+// NewTimelockConverter creates a new TimelockConverter
+func NewTimelockConverter() sdk.TimelockConverter {
+	return &timelockConverter{}
+}
+
+func (t *timelockConverter) ConvertBatchToChainOperations(
+	_ context.Context,
+	metadata types.ChainMetadata,
+	bop types.BatchOperation,
+	timelockAddress string,
+	mcmAddress string,
+	delay types.Duration,
+	action types.TimelockAction,
+	predecessor common.Hash,
+	salt common.Hash,
+) ([]types.Operation, common.Hash, error) {
+	// Create the list of RBACTimelockCall (batch of calls) and tags for the operations
+	calls, err := ConvertBatchToCalls(bop)
+	if err != nil {
+		return []types.Operation{}, common.Hash{}, fmt.Errorf("failed to convert batch to calls: %w", err)
+	}
+
+	tags := make([]string, 0)
+	for _, tx := range bop.Transactions {
+		tags = append(tags, tx.Tags...)
+	}
+
+	operationID, errHash := HashOperationBatch(calls, predecessor, salt)
+	if errHash != nil {
+		return []types.Operation{}, common.Hash{}, errHash
+	}
+
+	qID, err := tvm.RandomQueryID()
+	if err != nil {
+		return []types.Operation{}, common.Hash{}, fmt.Errorf("failed to generate random query ID: %w", err)
+	}
+
+	// Encode the data based on the operation
+	var data *cell.Cell
+	switch action {
+	case types.TimelockActionSchedule:
+		data, err = tlb.ToCell(timelock.ScheduleBatch{
+			QueryID: qID,
+
+			Calls:       calls,
+			Predecessor: tlbe.NewUint256(predecessor.Big()),
+			Salt:        tlbe.NewUint256(salt.Big()),
+			Delay:       uint32(delay.Seconds()),
+		})
+	case types.TimelockActionCancel:
+		data, err = tlb.ToCell(timelock.Cancel{
+			QueryID: qID,
+
+			ID: tlbe.NewUint256(operationID.Big()),
+		})
+	case types.TimelockActionBypass:
+		data, err = tlb.ToCell(timelock.BypasserExecuteBatch{
+			QueryID: qID,
+
+			Calls: calls,
+		})
+	default:
+		return []types.Operation{}, common.Hash{}, sdkerrors.NewInvalidTimelockOperationError(string(action))
+	}
+
+	if err != nil {
+		return []types.Operation{}, common.Hash{}, fmt.Errorf("failed to encode timelock action data: %w", err)
+	}
+
+	// Map to Ton Address type
+	dstAddr, err := address.ParseAddr(timelockAddress)
+	if err != nil {
+		return []types.Operation{}, common.Hash{}, fmt.Errorf("invalid timelock address: %w", err)
+	}
+
+	// Notice: follows EVM convention of value being 0 here
+	value := big.NewInt(0)
+
+	var tx types.Transaction
+	tx, err = NewTransaction(dstAddr, data.BeginParse(), value, "RBACTimelock", tags)
```

Comment:
shouldn't this contract type be fetched from the bop BatchOperation input? rather than hard coded as `RBACTimelock`, which is not supported by the TON decoder since it's missing from the `TLBsByContract`